File: packages/@aiengineeringharness/wocode/agent/skills/build-tool-skill/assets/compliance-check.py
#!/usr/bin/env python3
"""
Compliance checker for skill files across all 7 AI coding tools.
Migrated from skill-compliance-checker.

Usage:
    python3 compliance-check.py                        # Check all tools
    python3 compliance-check.py --tool=opencode         # Single tool
    python3 compliance-check.py --fix                   # Auto-fix issues
    python3 compliance-check.py --json                  # JSON output
"""

import os
import sys
import re
import json
import glob
import yaml

HARNESS_ROOT = os.path.join(os.path.dirname(__file__), "..", "..", "..", "..")
HARNESS_ROOT = os.path.abspath(HARNESS_ROOT)

# Per-tool spec as defined in AGENTS.md Naming Conventions table
TOOL_SPECS = {
    "opencode": {
        "naming": "kebab",
        "allowed_tools_case": "lowercase",
        "dir_prefix": "opencode/skills/",
    },
    "claude": {
        "naming": "snake",
        "allowed_tools_case": "pascalcase",
        "dir_prefix": "claude/skills/",
    },
    "gemini": {
        "naming": "snake",
        "allowed_tools_case": "lowercase",
        "dir_prefix": "gemini/skills/",
    },
    "pi": {
        "naming": "kebab",
        "allowed_tools_case": "titlecase",
        "dir_prefix": "pi/agent/skills/",
    },
    "wocode": {
        "naming": "snake",
        "allowed_tools_case": "lowercase",
        "dir_prefix": "wocode/agent/skills/",
    },
    "antigravity": {
        "naming": "snake",
        "allowed_tools_case": "lowercase",
        "dir_prefix": "antigravity/skills/",
    },
    "codex": {
        "naming": "snake",
        "allowed_tools_case": "lowercase_snake",
        "dir_prefix": "codex/skills/",
    },
}


def check_naming_convention(dir_name: str, spec: dict) -> list[str]:
    errors = []
    naming = spec["naming"]
    if naming == "kebab":
        if not re.match(r"^[a-z0-9]+(-[a-z0-9]+)*$", dir_name):
            errors.append(f"WRONG_NAMING_CONVENTION: '{dir_name}' should be kebab-case")
    elif naming == "snake":
        if not re.match(r"^[a-z0-9]+(_[a-z0-9]+)*$", dir_name):
            errors.append(f"WRONG_NAMING_CONVENTION: '{dir_name}' should be snake_case")
    return errors


def check_allowed_tools_case(tools: list[str], spec: dict) -> list[str]:
    errors = []
    expected_case = spec["allowed_tools_case"]
    for tool in tools:
        if expected_case == "lowercase" and tool != tool.lower():
            errors.append(f"WRONG_TOOL_CASE: '{tool}' should be lowercase")
        elif expected_case == "pascalcase" and tool != tool.capitalize():
            errors.append(f"WRONG_TOOL_CASE: '{tool}' should be PascalCase")
        elif expected_case == "titlecase" and tool != tool.title():
            errors.append(f"WRONG_TOOL_CASE: '{tool}' should be Title Case")
        elif expected_case == "lowercase_snake" and not re.match(r"^[a-z]+_[a-z]+$", tool):
            errors.append(f"WRONG_TOOL_CASE: '{tool}' should be lowercase_snake")
    return errors


def check_skill_file(skill_dir: str, tool: str, spec: dict) -> list[dict]:
    issues = []
    dir_name = os.path.basename(skill_dir)
    skill_path = os.path.join(skill_dir, "SKILL.md")

    if not os.path.exists(skill_path):
        issues.append({"file": skill_path, "code": "FILE_NOT_FOUND", "message": "SKILL.md not found"})
        return issues

    with open(skill_path) as f:
        content = f.read()

    # Parse frontmatter
    match = re.match(r"^---\n(.*?)\n---\n", content, re.DOTALL)
    if not match:
        issues.append({"file": skill_path, "code": "NO_FRONTMATTER", "message": "No YAML frontmatter found"})
        return issues

    try:
        fm = yaml.safe_load(match.group(1))
    except yaml.YAMLError as e:
        issues.append({"file": skill_path, "code": "YAML_ERROR", "message": str(e)})
        return issues

    if not isinstance(fm, dict):
        return issues

    # Check name matches directory
    name = fm.get("name", "")
    if name != dir_name:
        issues.append({
            "file": skill_path,
            "code": "NAME_MISMATCH",
            "message": f"name '{name}' != directory '{dir_name}'",
        })

    # Check naming convention
    naming_errors = check_naming_convention(dir_name, spec)
    for e in naming_errors:
        issues.append({"file": skill_path, "code": e.split(":")[0], "message": e})

    # Check allowed-tools casing
    tools = fm.get("allowed-tools", "")
    if isinstance(tools, str):
        tools = [t.strip() for t in tools.split(",") if t.strip()]
    if isinstance(tools, list):
        case_errors = check_allowed_tools_case(tools, spec)
        for e in case_errors:
            issues.append({"file": skill_path, "code": e.split(":")[0], "message": e})

    return issues


def main():
    args = sys.argv[1:]
    tool_filter = None
    fix_mode = "--fix" in args
    json_mode = "--json" in args

    for arg in args:
        if arg.startswith("--tool="):
            tool_filter = arg.split("=", 1)[1]

    tools_to_check = [tool_filter] if tool_filter else list(TOOL_SPECS.keys())
    all_issues = {}

    for tool in tools_to_check:
        if tool not in TOOL_SPECS:
            print(f"Unknown tool: {tool}")
            sys.exit(1)

        spec = TOOL_SPECS[tool]
        skills_glob = os.path.join(HARNESS_ROOT, spec["dir_prefix"], "*", "SKILL.md")
        skill_dirs = set(os.path.dirname(p) for p in glob.glob(skills_glob))

        tool_issues = []
        for skill_dir in sorted(skill_dirs):
            issues = check_skill_file(skill_dir, tool, spec)
            tool_issues.extend(issues)

        all_issues[tool] = tool_issues

    if json_mode:
        print(json.dumps(all_issues, indent=2))
    else:
        total = sum(len(v) for v in all_issues.values())
        for tool, issues in all_issues.items():
            status = "OK" if not issues else f"{len(issues)} issue(s)"
            print(f"{tool}: {status}")
            for issue in issues:
                print(f"  [{issue['code']}] {issue['file']}")
                print(f"    {issue['message']}")
        print(f"\nTotal: {total} issue(s) across {len(tools_to_check)} tool(s)")

    total_issues = sum(len(v) for v in all_issues.values())
    sys.exit(0 if total_issues == 0 else 1)


if __name__ == "__main__":
    main()

File: packages/@aiengineeringharness/scripts/compliance-check.ts
#!/usr/bin/env -S deno run -A
/**
 * Phase 4: Online Compliance Checking
 *
 * Reads skill SKILL.md files across all tools and validates them against
 * known tool specifications from docs/tools/ai-coding-tools/.
 *
 * Checks:
 *   1. Tool name usage (e.g., Pi using PascalCase tool names vs OpenCode lowercase)
 *   2. Frontmatter field validity per tool (e.g., `disable-model-invocation` unsupported in OpenCode)
 *   3. Known tool name mismatches in body text
 *   4. MCP disable syntax per tool
 *
 * Usage:
 *   deno run -A scripts/compliance-check.ts
 *   deno run -A scripts/compliance-check.ts --tool=opencode
 *   deno run -A scripts/compliance-check.ts --fix
 */

import { parseArgs } from "jsr:@std/cli@1/parse-args";
import { join } from "jsr:@std/path@1/join";
import { walk } from "jsr:@std/fs@1/walk";
import { parse as parseYaml } from "jsr:@std/yaml@1";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface ToolSpec {
  name: string; // tool key (opencode, claude, pi, etc.)
  naming: "snake" | "kebab";
  toolNameCase: "PascalCase" | "lowercase" | "UPPERCASE";
  allowedToolsFormat: "PascalCase" | "lowercase" | "space_separated_tools";
  supportedFrontmatter: string[];
  disableSyntax: string; // e.g., "enabled: false" for OpenCode, "disabled: true" for Claude
  knownToolNames: string[];
  deprecatedPatterns: string[];
  extraChecks: Record<string, string[]>;
}

interface ComplianceIssue {
  file: string;
  line: number;
  severity: "error" | "warning" | "info";
  code: string;
  message: string;
}

interface ComplianceResult {
  tool: string;
  skill: string;
  issues: ComplianceIssue[];
  passed: boolean;
}

// ---------------------------------------------------------------------------
// Tool specifications (compiled from docs/tools/ai-coding-tools/)
// ---------------------------------------------------------------------------

const TOOL_SPECS: Record<string, ToolSpec> = {
  opencode: {
    name: "opencode",
    naming: "snake",
    toolNameCase: "lowercase",
    allowedToolsFormat: "lowercase",
    supportedFrontmatter: ["name", "description", "allowed-tools", "docs-url", "disable-model-invocation", "on"],
    disableSyntax: '"enabled": false',
    knownToolNames: ["bash", "edit", "write", "read", "grep", "glob", "lsp", "apply_patch", "skill", "todowrite", "webfetch", "websearch", "question"],
    deprecatedPatterns: [],
    extraChecks: {},
  },
  claude: {
    name: "claude",
    naming: "snake",
    toolNameCase: "PascalCase",
    allowedToolsFormat: "PascalCase",
    supportedFrontmatter: ["name", "description", "allowed-tools", "disable-model-invocation"],
    disableSyntax: '"disabled": true',
    knownToolNames: ["Agent", "AskUserQuestion", "Bash", "CronCreate", "CronDelete", "CronList", "Edit", "EnterPlanMode", "EnterWorktree", "ExitPlanMode", "ExitWorktree", "Glob", "Grep", "ListMcpResourcesTool", "LSP", "Monitor", "NotebookEdit", "PowerShell", "PushNotification", "Read", "ReadMcpResourceTool", "RemoteTrigger", "ScheduleWakeup", "SendMessage", "ShareOnboardingGuide", "Skill", "TaskCreate", "TaskGet", "TaskList", "TaskStop", "TaskUpdate", "TeamCreate", "TeamDelete", "TodoWrite", "ToolSearch", "WaitForMcpServers", "WebFetch", "WebSearch", "Workflow", "Write"],
    deprecatedPatterns: ["TodoWrite"],
    extraChecks: {},
  },
  gemini: {
    name: "gemini",
    naming: "snake",
    toolNameCase: "lowercase",
    allowedToolsFormat: "snake_case",
    supportedFrontmatter: ["name", "description", "allowed-tools", "docs-url", "on"],
    disableSyntax: "N/A",
    knownToolNames: ["read_file", "run_shell_command", "edit_file", "write_file", "glob", "grep", "web_fetch", "web_search", "ask_user", "task"],
    deprecatedPatterns: [],
    extraChecks: {},
  },
  pi: {
    name: "pi",
    naming: "kebab",
    toolNameCase: "PascalCase",
    allowedToolsFormat: "word", // space-separated PascalCase
    supportedFrontmatter: ["name", "description", "allowed-tools"],
    disableSyntax: "N/A",
    knownToolNames: ["Read", "Bash", "Edit", "Write", "Grep", "Glob", "WebFetch", "WebSearch", "Question", "Skill", "Task", "LSP"],
    deprecatedPatterns: [],
    extraChecks: {},
  },
  wocoder: {
    name: "wocoder",
    naming: "snake",
    toolNameCase: "lowercase",
    allowedToolsFormat: "lowercase",
    supportedFrontmatter: ["name", "description", "allowed-tools", "docs-url", "disable-model-invocation", "on"],
    disableSyntax: '"enabled": false',
    knownToolNames: ["bash", "edit", "write", "read", "grep", "glob", "skill", "todowrite", "webfetch", "websearch", "question"],
    deprecatedPatterns: [],
    extraChecks: {},
  },
  codex: {
    name: "codex",
    naming: "snake",
    toolNameCase: "lowercase",
    allowedToolsFormat: "lowercase",
    supportedFrontmatter: ["name", "description", "allowed-tools"],
    disableSyntax: "N/A",
    knownToolNames: ["bash", "edit", "write", "read", "grep", "glob", "web_fetch", "web_search"],
    deprecatedPatterns: [],
    extraChecks: {},
  },
  antigravity: {
    name: "antigravity",
    naming: "snake",
    toolNameCase: "lowercase",
    allowedToolsFormat: "lowercase",
    supportedFrontmatter: ["name", "description", "allowed-tools", "docs-url", "disable-model-invocation", "on"],
    disableSyntax: '"enabled": false',
    knownToolNames: ["bash", "edit", "write", "read", "grep", "glob", "skill", "todowrite", "webfetch", "websearch", "question"],
    deprecatedPatterns: [],
    extraChecks: {},
  },
};

const REPO_ROOT = join(import.meta.dirname!, "..", "..", "..");

// ---------------------------------------------------------------------------
// Parsing helpers
// ---------------------------------------------------------------------------

function parseFrontmatter(content: string): { frontmatter: Record<string, unknown>; body: string; error?: string } {
  const match = content.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  if (!match) {
    return { frontmatter: {}, body: content, error: "No valid frontmatter found" };
  }
  try {
    const fm = parseYaml(match[1]) as Record<string, unknown>;
    return { frontmatter: fm ?? {}, body: match[2] };
  } catch (e) {
    return { frontmatter: {}, body: match[2], error: `YAML parse error: ${e}` };
  }
}

function readLines(content: string): string[] {
  return content.split("\n");
}

// ---------------------------------------------------------------------------
// Compliance checks
// ---------------------------------------------------------------------------

function checkFrontmatterFields(
  skillName: string,
  frontmatter: Record<string, unknown>,
  spec: ToolSpec,
): ComplianceIssue[] {
  const issues: ComplianceIssue[] = [];
  const validFields = spec.supportedFrontmatter;

  for (const key of Object.keys(frontmatter)) {
    if (key === "name" || key === "description") continue; // always valid
    if (key === "allowed-tools" && !spec.allowedToolsFormat) continue;
    if (!validFields.includes(key)) {
      issues.push({
        file: skillName,
        line: 0,
        severity: "warning",
        code: "UNSUPPORTED_FRONTMATTER",
        message: `Frontmatter field "${key}" is not in supported fields for ${spec.name}: ${validFields.join(", ")}`,
      });
    }
  }

  return issues;
}

function checkToolNameCase(
  skillName: string,
  frontmatter: Record<string, unknown>,
  spec: ToolSpec,
): ComplianceIssue[] {
  const issues: ComplianceIssue[] = [];
  const allowed = frontmatter["allowed-tools"];
  if (!allowed || typeof allowed !== "string") return issues;

  const names = allowed.split(/\s+/).filter(Boolean);
  for (const name of names) {
    // Check if name matches expected case
    if (spec.toolNameCase === "PascalCase") {
      if (!/^[A-Z][a-z]/.test(name) && name.length > 1) {
        issues.push({
          file: skillName,
          line: 0,
          severity: "error",
          code: "WRONG_TOOL_CASE",
          message: `Tool name "${name}" should be PascalCase for ${spec.name} (e.g., "Read" not "read")`,
        });
      }
    } else if (spec.toolNameCase === "lowercase") {
      if (name !== name.toLowerCase()) {
        issues.push({
          file: skillName,
          line: 0,
          severity: "error",
          code: "WRONG_TOOL_CASE",
          message: `Tool name "${name}" should be lowercase for ${spec.name} (e.g., "read" not "Read")`,
        });
      }
    }
  }

  return issues;
}

function checkMentionedToolNames(
  skillName: string,
  content: string,
  spec: ToolSpec,
): ComplianceIssue[] {
  const issues: ComplianceIssue[] = [];

  // Skip the frontmatter
  const body = content.replace(/^---[\s\S]*?---\n/, "");

  // Check for wrong-case tool names in body (e.g., Pi body mentioning "Read" is fine,
  // but OpenCode body mentioning "Read" instead of "read" is wrong)
  if (spec.toolNameCase === "lowercase") {
    const lines = readLines(body);
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const pascalMatches = line.match(/\b(Read|Write|Edit|Bash|Grep|Glob|WebFetch|WebSearch|AskUserQuestion|Skill)\b/g);
      if (pascalMatches) {
        issues.push({
          file: skillName,
          line: i + 1,
          severity: "warning",
          code: "BODY_WRONG_TOOL_CASE",
          message: `PascalCase tool name "${pascalMatches[0]}" in body — should be lowercase for ${spec.name}`,
        });
      }
    }
  } else if (spec.toolNameCase === "PascalCase") {
    const lines = readLines(body);
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const lowerMatches = line.match(/\b(read|write|edit|bash|grep|glob|webfetch|websearch)\b/g);
      if (lowerMatches) {
        // Only flag if it's clearly a tool reference (not a regular word)
        const word = lowerMatches[0];
        if (["read", "write", "edit", "bash", "grep", "glob"].includes(word)) {
          issues.push({
            file: skillName,
            line: i + 1,
            severity: "warning",
            code: "BODY_WRONG_TOOL_CASE",
            message: `Lowercase tool name "${word}" in body — should be PascalCase for ${spec.name}`,
          });
        }
      }
    }
  }

  return issues;
}

function checkDeprecatedPatterns(
  skillName: string,
  content: string,
  spec: ToolSpec,
): ComplianceIssue[] {
  const issues: ComplianceIssue[] = [];

  for (const pattern of spec.deprecatedPatterns) {
    if (content.includes(pattern)) {
      issues.push({
        file: skillName,
        line: 0,
        severity: "warning",
        code: "DEPRECATED_PATTERN",
        message: `Deprecated pattern "${pattern}" found — should be avoided for ${spec.name}`,
      });
    }
  }

  return issues;
}

function checkNamingConvention(
  skillName: string,
  spec: ToolSpec,
): ComplianceIssue[] {
  const issues: ComplianceIssue[] = [];

  if (spec.naming === "kebab" && skillName.includes("_")) {
    issues.push({
      file: skillName,
      line: 0,
      severity: "error",
      code: "WRONG_NAMING_CONVENTION",
      message: `Skill dir "${skillName}" uses snake_case but ${spec.name} requires kebab-case`,
    });
  } else if (spec.naming === "snake" && skillName.includes("-")) {
    issues.push({
      file: skillName,
      line: 0,
      severity: "error",
      code: "WRONG_NAMING_CONVENTION",
      message: `Skill dir "${skillName}" uses kebab-case but ${spec.name} requires snake_case`,
    });
  }

  return issues;
}

function checkFrontmatterNameMatch(
  skillDir: string,
  frontmatter: Record<string, unknown>,
  spec: ToolSpec,
): ComplianceIssue[] {
  const issues: ComplianceIssue[] = [];
  const fmName = frontmatter["name"];
  if (fmName && typeof fmName === "string") {
    const expectedName = spec.naming === "kebab"
      ? skillDir.replace(/_/g, "-")
      : skillDir.replace(/-/g, "_");
    if (fmName !== expectedName) {
      issues.push({
        file: skillDir,
        line: 0,
        severity: "warning",
        code: "NAME_MISMATCH",
        message: `Frontmatter name "${fmName}" doesn't match dir name "${skillDir}" (expected "${expectedName}")`,
      });
    }
  }
  return issues;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function checkTool(toolName: string, fixMode: boolean): Promise<ComplianceResult[]> {
  const spec = TOOL_SPECS[toolName];
  if (!spec) {
    console.error(`Unknown tool: ${toolName}`);
    Deno.exit(1);
  }

  const skillDir = join(REPO_ROOT, "packages/@aiengineeringharness", toolName, "skills");
  const results: ComplianceResult[] = [];

  try {
    const entries = Deno.readDirSync(skillDir);
    for (const entry of entries) {
      if (!entry.isDirectory) continue;
      const skillPath = join(skillDir, entry.name, "SKILL.md");

      let content: string;
      try {
        content = Deno.readTextFileSync(skillPath);
      } catch {
        results.push({
          tool: toolName,
          skill: entry.name,
          issues: [{ file: entry.name, line: 0, severity: "error", code: "MISSING_SKILL", message: "SKILL.md not found" }],
          passed: false,
        });
        continue;
      }

      const { frontmatter, body, error } = parseFrontmatter(content);
      const issues: ComplianceIssue[] = [];

      if (error) {
        issues.push({ file: entry.name, line: 0, severity: "error", code: "PARSE_ERROR", message: error });
      }

      issues.push(...checkNamingConvention(entry.name, spec));
      issues.push(...checkFrontmatterFields(entry.name, frontmatter, spec));
      issues.push(...checkToolNameCase(entry.name, frontmatter, spec));
      issues.push(...checkMentionedToolNames(entry.name, content, spec));
      issues.push(...checkDeprecatedPatterns(entry.name, content, spec));
      issues.push(...checkFrontmatterNameMatch(entry.name, frontmatter, spec));

      results.push({
        tool: toolName,
        skill: entry.name,
        issues,
        passed: issues.length === 0,
      });
    }
  } catch (e) {
    console.error(`Error reading ${skillDir}: ${e}`);
  }

  return results;
}

function printResults(results: ComplianceResult[]): void {
  const bySeverity: Record<string, ComplianceIssue[]> = { error: [], warning: [], info: [] };

  for (const result of results) {
    for (const issue of result.issues) {
      bySeverity[issue.severity]?.push(issue);
    }
  }

  console.log("\n── Compliance Report ──\n");

  const total = results.length;
  const passed = results.filter((r) => r.passed).length;
  const failed = results.filter((r) => !r.passed).length;

  for (const [severity, issues] of Object.entries(bySeverity)) {
    if (issues.length === 0) continue;
    const icon = severity === "error" ? "✗" : severity === "warning" ? "△" : "ℹ";
    console.log(`  ${icon} ${severity.toUpperCase()}: ${issues.length}`);
    for (const issue of issues) {
      const loc = issue.line > 0 ? `:${issue.line}` : "";
      console.log(`      [${issue.code}] ${issue.tool}/${issue.file}${loc} — ${issue.message}`);
    }
    console.log();
  }

  console.log(`  Skills: ${total} total, ${passed} passed, ${failed} with issues`);
  console.log();
}

async function main(): Promise<void> {
  const args = parseArgs(Deno.args, {
    string: ["tool"],
    boolean: ["fix"],
  });

  const allTools = Object.keys(TOOL_SPECS);
  const toolFilter = args.tool ? [args.tool] : allTools;
  const fixMode = !!args.fix;

  console.log(`\nPhase 4: Online Compliance Check${fixMode ? " (fix mode)" : ""}`);
  console.log(`Checking tools: ${toolFilter.join(", ")}`);

  for (const tool of toolFilter) {
    const results = await checkTool(tool, fixMode);
    printResults(results);
  }
}

if (import.meta.main) {
  await main();
}
